### ISSUE TITLE
Widevine installation silently fails (or takes a very long time)

### ISSUE THREAD

**Issue by koenoe** [castlabs/electron-releases#62]
Initially we got a lot of reports of users who were intermittently not getting the BrowserWindow on startup (we only create this on `widevine-ready`). But logs also didn't show any errors coming in on `widevine-error`. Therefore I created a few helpers and tried to make our implementation a bit more robust so it would give us more and hopefully new insights.

Our helpers:
```
export function onAppReady(cb: Function) {
  if (app.isReady()) {
    cb();
  } else {
    app.on('ready', () => cb());
  }
}

const WIDEVINE_TIMEOUT = 30000;
let widevineReady: boolean;
let widevineError: any;

export function onWidevineReady(success: Function, error: Function) {
  if (widevineReady) {
    logger.info('Widevine already installed');
    return success();
  }

  if (widevineError) {
    logger.info(`Widevine already errored: ${widevineError.message}`);
    return error(widevineError);
  }

  const timeoutId = setTimeout(() => {
    logger.crit('Widevine timed out.');
    widevineError = new Error('Widevine timed out.');
    error(widevineError);
  }, WIDEVINE_TIMEOUT);

  app.on('widevine-error', (err: Error) => {
    logger.crit(`Widevine error: ${err.message}`);
    widevineReady = true; // this should be widevineError = err eventually;
    clearTimeout(timeoutId);
    success(); // this should be error() eventually, but for some users Widevine installation always fails and are stuck on DRM error screen
  });

  app.on('widevine-ready', (version, lastVersion) => {
    if (lastVersion !== null) {
      logger.info('Widevine upgraded', { version, lastVersion });
    } else {
      logger.info('Widevine installed', { version });
    }
    widevineReady = true;
    clearTimeout(timeoutId);
    success();
  });

  app.on('widevine-update-pending', (curVersion, pendingVersion) => {
    logger.info('New Widevine update pending', {
      currentVersion: curVersion,
      pendingVersion,
    });
  });
}
```

Our `main.js`
```
function appReady(): Promise<void> {
  return new Promise((resolve) => {
    onAppReady(() => resolve());
  });
}

function widevineReady(): Promise<void> {
  return new Promise((resolve, reject) => {
    onWidevineReady(resolve, reject);
  });
}

await appReady();
try {
  // Note: We _must_ wait for Widevine to be installed before starting the
  // web. We have a separate client token if Widevine exists. If Widevine was
  // installed after this check, we wouldn't get DRM content.
  await widevineReady();

  this.windowController.createWindow();
} catch (error) {
  this.windowController.showWidevineErrorPage();
}
```

Basically what we're trying to achieve is that we show the user an error page with a button that restarts the app when Widevine installation takes longer than 30 seconds. It looks like when users get this page, Widevine always seems to successfully install after restarting the app by pressing the button.

TL;DR
How long can it take before `widevine-ready` is fired? Should we increase our timeout safety net or do you recommend a different approach? Could Widevine installation silently fail?

We're on Electron 9.1.2 btw, but we've had the same issues in Electron 7 and 8.

Thanks in advance.

Koen

**Comment by khwaaj** [comment]
Hi Koen,

There should not be a code path where one of `widevine-ready` or `widevine-error` is not fired, as long as `no-verify-widevine-cdm` has not been set. That said, I can see a couple of reasons this might happen:

1. The CDM download hangs, or takes a indeterminate amount of time, for whatever reason. The download is only a few megabyts so it would not take much time under normal conditions, but in cases where the network is slow, or somehow compromised, it might take more time.

2. We have an issue somewhere and are silently letting an error slip through without notifying.

I've not actually seen anything like this happen myself so I'm trying to figure out how to best go about finding the culprit. One thing that is not entirely clear to me is if this occurs only on first start, i.e. when no CDM is previously installed, or if it happens when looking for a CDM update? Or both? Also, are you seeing the same behaviour across platforms?

About the timeout, it is a good idea since it will, at least, give some feedback to the user in a timely fashion. 30s should be plenty under normal circumstances, but it is a bit difficult to know how long the download might take under extraordinary conditions, so the value becomes a bit arbitrary at that point. I guess the most reasonable approach is to choose a value based on when you'd expect the user to get impatient waiting for the window to show up.

The best thing, of course, would be if we can identify the culprit behind this and handle it better in our code. From the above I suppose you have no reliable way to reproduce this, but if you have some ideas that might help, I'm all ears.

Emil

**Comment by koenoe** [comment]
> One thing that is not entirely clear to me is if this occurs only on first start, i.e. when no CDM is previously installed, or if it happens when looking for a CDM update? Or both?

It seems to happen on both. It does look like it happens mostly on a [restart after an app update](https://www.electronjs.org/docs/api/auto-updater#autoupdaterquitandinstall). 

> Also, are you seeing the same behaviour across platforms?
Initially I thought I had found something as we only offer 32-bit builds for Windows, but then I also found quite some events in Sentry from Mac users. It does look like to happen more often on Windows. Do you think a 32-bit build for users with a 64-bit architecture could cause issues with Widevine?

Unfortunately I don't have steps to reproduce, but as we speak I'm trying to consistently reproduce it locally as I just had it happening locally too (on my Windows 64-bit machine).

**Comment by khwaaj** [comment]
> It seems to happen on both. It does look like it happens mostly on a [restart after an app update](https://www.electronjs.org/docs/api/auto-updater#autoupdaterquitandinstall) though.

Ok, so it is likely not anything to do with the actual download then, as an update won't wait for the download to complete before posting `widevine-ready` since there is already a CDM in play. That narrows it down a bit.

> Initially I thought I had found something as we only offer 32-bit builds for Windows, but then I also found quite some events in Sentry from Mac users. It does look like to happen more often on Windows. Do you think a 32-bit build for users with a 64-bit architecture could cause issues with Widevine?

No, that's unlikely. I've seen no indication that a 32-bit build on 64-bit Windows would cause any issues with Widevine.

> Unfortunately I don't have steps to reproduce, but as we speak I'm trying to consistently reproduce it as I just had it happening locally too (on my Windows 64-bit machine).

Ok, I'll try to set up some testing of my own based on the information you provided. If you find a way to reproduce it, maybe we can collaborate and see if we can't get this fixed permanently.

**Comment by khwaaj** [comment]
Btw, you are not seeing any UnhandledPromiseRejectionWarning errors in the logs when this is happening? I'm thinking there could be some code path that has a faulty promise rejection that leaves a new orphaned promise instead of rejecting the correct one.

**Comment by koenoe** [comment]
I was able to consistently reproduce locally when changing our app version. We have some logic in place to clear the service workers by calling [clearStorageData](https://www.electronjs.org/docs/api/session#sesclearstoragedataoptions) on the default session. Seems like this Promise never got resolved and somehow messed up our flow. I've put a fix out for this, so hopefully we won't have users reporting the issue anymore after the next release.

If somehow this wasn't the issue, I'll re-open this issue.

Many thanks for your help!
